src/lookup.rs

```diff
@@ -436,6 +474,101 @@ pub(super) async fn fetch_channel_updates<L: Deref>(delta_set: &mut DeltaSet, cl
+	let last_sync_timestamp_float = last_sync_timestamp as f64;
+
+	// get the latest node updates prior to last_sync_timestamp
+	let reference_rows = client.query_raw("
+		SELECT DISTINCT ON (public_key) public_key, CAST(EXTRACT('epoch' from seen) AS BIGINT) AS seen, announcement_signed FROM node_announcements
+		WHERE seen < TO_TIMESTAMP($1)
+		ORDER BY public_key ASC, seen DESC
+		", [last_sync_timestamp_float]).await.unwrap();
+	let mut pinned_rows = Box::pin(reference_rows);
+
+	log_info!(logger, "Fetched node announcement reference rows in {:?}", start.elapsed());
+
+	let mut reference_row_count = 0;
+
+	while let Some(row_res) = pinned_rows.next().await {
+		let current_reference = row_res.unwrap();
+
+		let seen = current_reference.get::<_, i64>("seen") as u32;
+		let blob: Vec<u8> = current_reference.get("announcement_signed");
+		let mut readable = Cursor::new(blob);
+		let unsigned_node_announcement = NodeAnnouncement::read(&mut readable).unwrap().contents;
+		let node_id = unsigned_node_announcement.node_id;
+
+		let current_node_delta = delta_set.entry(node_id).or_insert(NodeDelta::default());
+		let address_set: HashSet<SocketAddress> = unsigned_node_announcement.addresses.into_iter().collect();
+		(*current_node_delta).last_details_before_seen.get_or_insert(NodeDetails {
+			seen,
+			features: unsigned_node_announcement.features,
+			addresses: address_set
+		});
+		log_gossip!(logger, "Node {} last update before seen: {} (seen at {})", node_id, unsigned_node_announcement.timestamp, seen);
+
+		reference_row_count += 1;
+	}
+
+
+	log_info!(logger, "Processed {} node announcement reference rows (delta size: {}) in {:?}",
+		reference_row_count, delta_set.len(), start.elapsed());
+
+	// get all the intermediate node updates
+	// (to calculate the set of mutated fields for snapshotting, where intermediate updates may
+	// have been omitted)
+	let intermediate_updates = client.query_raw("
+		SELECT announcement_signed, CAST(EXTRACT('epoch' from seen) AS BIGINT) AS seen
+		FROM node_announcements
+		WHERE seen >= TO_TIMESTAMP($1)
+		ORDER BY public_key ASC, timestamp DESC
+		", [last_sync_timestamp_float]).await.unwrap();
+	let mut pinned_updates = Box::pin(intermediate_updates);
+	log_info!(logger, "Fetched intermediate node announcement rows in {:?}", start.elapsed());
+
+	let mut previous_node_id: Option<NodeId> = None;
+
+	let mut intermediate_update_count = 0;
+	while let Some(row_res) = pinned_updates.next().await {
+		let intermediate_update = row_res.unwrap();
+		intermediate_update_count += 1;
+
+		let current_seen_timestamp = intermediate_update.get::<_, i64>("seen") as u32;
+		let blob: Vec<u8> = intermediate_update.get("announcement_signed");
+		let mut readable = Cursor::new(blob);
+		let unsigned_node_announcement = NodeAnnouncement::read(&mut readable).unwrap().contents;
+
+		let node_id = unsigned_node_announcement.node_id;
+		let is_previously_processed_node_id = Some(node_id) == previous_node_id;
```

Comment:
Can't we skip the rest of the loop iteration if this is true (other than updating `previous_node_id`) since we are iterating by descending timestamp? Could we also use `SELECT DISTINCT` then?

Reply:
Could you add a comment why this is the case? It isn't readily obvious.